BUG: core: fix wrong method flags for scalartypes.c.src:gentype_copy

pv opened the pull request:
The method declaration is
```
/*
 * These gentype_* functions take keyword arguments.
 * The proper flag is METH_VARARGS | METH_KEYWORDS.
 */
/**begin repeat
 *
 * #name = take, getfield, put, repeat, tofile, mean, trace, diagonal, clip,
 *         std, var, sum, cumsum, prod, cumprod, compress, sort, argsort,
 *         round, argmax, argmin, max, min, ptp, any, all, astype, resize,
 *         reshape, choose, tostring, tobytes, copy, searchsorted, view,
 *         flatten, ravel#
 */
static PyObject *
gentype_@name@(PyObject *self, PyObject *args, PyObject *kwds)
{
    return gentype_generic_method(self, args, kwds, "@name@");
}
/**end repeat**/
```
so it has to have `METH_KEYWORDS`.

Fixes gh-9608

pv committed:
BUG: core: fix wrong method flags for scalartypes.c.src:gentype_copy

charris commented:
Hmm, ndarray handles this a bit differently by defining `array_copy_keeporder` in `methods.c` that calls the regular copy method, but the signature of that function looks incorrect for `METH_VARARGS`, so that might need a fix.  I suppose a similar solution would work for scalars as `copy.copy` doesn't pass keywords, but it might not be worth the effort as order doesn't matter for scalars.

pv committed:
BUG: core: fix array_copy_keeporder METH_VARARGS signature

pv commented:
Fixed also array_copy_keeporder

charris commented:
The `gentype__deepcopy__` function is already correct, so different fix would be to simply move `__copy__` into the same generation loop.

pv commented:
I won't have time to return to this for a week+

charris commented:
@pv I might pull it down and change the fix if you are OK with that.

pv commented:
That's fine, see also https://help.github.com/articles/committing-chang
es-to-a-pull-request-branch-created-from-a-fork/

charris committed:
MAINT: Different fix for scalartypes.c.src `__copy__` method.

Create a new function `scalartype.c.src:gentype___copy__` that does not
take keyword arguments. This is done by adding `__copy__` to the same
code generator loop that generates `gentype___deepcopy__`.

charris commented:
Thanks Pauli.

mattip commented:
the test added in this pull request hits a ``RuntimeError('maximum recursion depth exceeded',)`` when printing ``np.void("a")`` in ``build_err_msg``, called from ``testing.assert_equal``. See issue #9345. Luckily for CPython that is an exception that can be caught, on PyPy it crashes the interpreter. Just a heads up, it is probably a PyPy problem but I need an issue to hang it on
